agent/agent_endpoint.go

```diff
@@ -1218,7 +1223,7 @@ func (s *HTTPHandlers) AgentRegisterService(resp http.ResponseWriter, req *http.
 }
 
 func (s *HTTPHandlers) AgentDeregisterService(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
-	serviceID := strings.TrimPrefix(req.URL.Path, "/v1/agent/service/deregister/")
```

Comment:
Edit: ah... it seems like we're still actually passing it to Consul's HTTP server via path parameter, but we're moving the location that we extract it from the URL path from `AgentDeregisterService` to `ServeHTTP`. Why does that change the behavior?

Reply:
Moving the extraction of service id from `AgentDeregisterService` to `ServeHTTP` solves the following problem:

- If the service id to be deregistered is an URL, e.g., `/v1/agent/service/deregister/http://example-service.com`,  the http handler (`w.handler.ServeHTTP(resp, req)`) will return 301 before passing to the `AgentDeregisterService` method. 

**Changed behavior**: so this PR removes the service id from the URL to allow ServeHTTP to continue passing the request to `AgentDeregisterService`.

**Backward compatible**: the change is backward compatible because there is no change to the external-facing API. In `ServeHTTP`, we strip the service from the URL, add it as query parameter, and pass to `AgentDeregisterService` (Please refer to the change in `http.go`)